src/riak_object.erl

```diff
@@ -623,12 +623,37 @@ get_value(Object=#r_object{}) ->
     [{_M,Value}] = get_contents(Object),
     Value.
 
-%% @doc calculates the hash of a riak object
--spec hash(riak_object()) -> integer().
+%% @doc calculates the canonical hash of a riak object
+%%      Old API which uses the version .
+-spec hash(riak_object()) -> binary().
 hash(Obj=#r_object{}) ->
-    Vclock = vclock(Obj),
-    UpdObj = riak_object:set_vclock(Obj, lists:sort(Vclock)),
-    erlang:phash2(to_binary(v0, UpdObj)).
+    %% BRIAN What to do here needs more discussion(race condition due to tick, vnode startup, etc)
+    case riak_kv_entropy_manager:get_version() of
+        0 ->
+            vclock_hash(Obj);
+        undefined ->
+            legacy_hash(Obj)
+    end.
+
+%% @doc calculates the canonical hash of a riak object depending on version
+-spec hash(riak_object(), non_neg_integer() | undefined) -> binary().
+hash(Obj=#r_object{}, _Version=0) ->
+    vclock_hash(Obj);
+hash(Obj=#r_object{}, _Version) ->
+    legacy_hash(Obj).
+
+%% @private return the legacy full object hash of the riak_object
+-spec legacy_hash(riak_object()) -> binary().
+legacy_hash(Obj=#r_object{}) ->
+    UpdObj = riak_object:set_vclock(Obj, lists:sort(vclock(Obj))),
+    Hash = erlang:phash2(to_binary(v0, UpdObj)),
+    term_to_binary(Hash).
+
+%% @private return the hash of the vclock
+-spec vclock_hash(riak_object()) -> binary().
+vclock_hash(Obj=#r_object{}) ->
+    Hash = erlang:phash2(to_binary(v0, lists:sort(vclock(Obj)))),
+    term_to_binary(Hash).
```

Comment:
As long as we're changing the hashing format, could we take this opportunity to optimize our hashing code any further?

For examples, it looks like `erlang:phash2` accepts arbitrary terms, so perhaps we don't need to wrap the result of `lists:sort(vclock(Obj))` in a call to `to_binary` - not sure if this would perform better or worse, but it seems like there's a good chance it would perform better since we'd be taking an extra encoding step out of the process.

Also, do we need to use `term_to_binary` at the end, or could we just do something more lightweight to convert it to a binary like `<<Hash:32>>`? That would almost certainly be faster, and yields smaller binaries too.